File: 1734-decode-xored-permutation/1734-decode-xored-permutation.py
class Solution:
    def fun(self, n):
        if n%4 == 1: return 1
        if n%4 == 2: return n+1
        if n%4 == 3: return 0
        return n
    
    def decode(self, encoded: List[int]) -> List[int]:
        n = len(encoded)+1;
        xor = self.fun(n);
        
        temp = 0;
        for i in range(0, n-1, 2):
            temp ^= encoded[i]
        
        last = xor ^ temp
        ans = [0]*n
        ans[n-1] = last
        for i in range(n-2, -1, -1):
            ans[i] = ans[i+1]^encoded[i]
        
        return ans
        

File: 1734-decode-xored-permutation/1734-decode-xored-permutation.java
class Solution {
    int fun(int n)
    {
        if(n%4 == 1) return 1;
        if(n%4 == 2) return n+1;
        if(n%4 == 3) return 0;
        return n;
    }
    
    public int[] decode(int[] encoded) {
        int n = encoded.length+1;
        int xor = fun(n);
        
        int temp = 0;
        for(int i=0; i<n-1; i+=2)
        {
            temp ^= encoded[i];
        }
        int last = xor ^ temp;
        int ans[] = new int[n];
        ans[n-1] = last;
        for(int i=n-2; i>=0; i--)
        {
            ans[i] = ans[i+1]^encoded[i];
        }
        return ans;
    }
}